devito/at_controller.py

```diff
@@ -0,0 +1,344 @@
+import os
+import stat
+import math
+import shutil
+import cpuinfo
+import subprocess
+
+# global vars.
+# report folder will be created in current dir. Change if necessary
+reports_folder_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), "auto-tune-reports")
+
+# where best block sizes for each kernel will be kept
+final_report_path = os.path.join(reports_folder_path, "final_report.txt")
+
+
+def get_at_block_size(time_order, space_order):
+    """Gets the best block sizes for given kernel if they are in the final_report_path
+
+    Args:
+        time_order (int): time order of kernel
+        space_order (int): space order of kernel
+
+    Returns:
+        list: best block size. Starting from outer most dimension
+        None: if report does not exist or does not contain required block sizes
+    """
+    global final_report_path
+
+    if not os.path.isfile(final_report_path):
+        return None  # returns none if report for best block sizes does not exist
+
+    with open(final_report_path, 'r') as f:
+        for line in f.readlines():
+
+            if "time" in line or "space" in line:  # ignores the first line of the report file
+                continue
+
+            split = line.split(' ')
+            if int(split[0]) == time_order and int(split[1]) == space_order:  # finds the one we are looking for
+
+                # Splits, converts all block sizes into int and returns
+                return [int(element) for element in split[2].split(',')]
+
+    return None  # returns none if no matching time/space order was found
+
+
+def get_optimal_block_size(shape, time_order, space_order):
+    """Gets optimal block size. Based on architecture Currently works only on linux
+
+    Args:
+        shape (tuple|list): shape of kernel
+        time_order (int): time order of kernel
+        space_order (int): space order of kernel
+
+    Returns:
+        int: most optimal size  for the block
+    """
+
+    # list [time_order][space_order] hardcoded number of loads from the roof line at: (change if necessary)
+    # https://docs.google.com/spreadsheets/d/1OmvsTftH3uCfYZj-1Lb-5Ji7edrURf0UzzywYaw0-FY/edit?ts=5745964f#gid=0
+    number_of_loads = [[11, 17, 23, 29, 35, 41, 47, 53], [19, 25, 31, 37, 43, 49, 55]]  # works only till space order 16
```

Comment:
I don't see how "At code should work as long as the code compiles as its built arround the main functionality" answers my concern. What are the cases where you see the propagator value being missing?